### PR TITLE
[Homework #18] - React UI by Usmonbek Rustamov

### PR DESCRIPTION
**Pull request by usmonbek-umarbekovich**
Hey, I completed the homework.

I haven't done a lof of projects using Typescript before. After the Typescript lesson, I really started to enjoy working with it.
Please, have a look at my homework and tell me if there is any mistakes or bad practices.
Thank you

### CODE REVIEW

File: homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  };
+
+  return (
+    <section className="avatar-section">
+      <ul className="avatar-list">
+        {avatarList.map(avatar => (
+          <Avatar
+            key={avatar.id}
+            id={avatar.id}
+            first_name={avatar.first_name}
```
Comment (VitalikMelnyk) [inline homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx:66]: I recommend writing props with camelCase.
E.g: first_name -> firstName and so on.

File: homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        });
+      });
+    } catch (error) {
+      console.log(error);
+    } finally {
+      setLoadingAvatarsId([]);
+    }
+  };
+
+  const handleRefreshAllAvatars = async () => {
```
Comment (VitalikMelnyk) [inline homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx:45]: You need to catch errors. Please review cases when you get errors from the backend and show some feedback for the user without crashing the application.
For example: when I added 20 avatars and then wanted to add 21 avatar I got the error.
Screenshot 2023-09-26 at 13 13 11

File: homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            first_name={avatar.first_name}
+            last_name={avatar.last_name}
+            url={avatar.url}
+            isLoading={loadingAvatarsId.includes(avatar.id)}
+            onRefreshAvatar={handleRefreshAvatar}
+          />
+        ))}
+        <li className="avatar">
+          {loadingAvatarsId.includes(-1) ? (
+            <div className="loader">
```
Comment (VitalikMelnyk) [inline homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx:75]: You can move this code to separate components more easily and conveniently by dividing them into small components.

File: homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+              <img
+                className="loader__image"
+                src={refreshImg}
+                alt="New avatar loading"
+              />
+            </div>
+          ) : (
+            <button className="btn btn-square" onClick={handleAddAvatar}>
```
Comment (VitalikMelnyk) [inline homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx:83]: The same here.

File: homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+              +
+            </button>
+          )}
+        </li>
+      </ul>
+
+      <button className="btn btn-refresh" onClick={handleRefreshAllAvatars}>
```
Comment (VitalikMelnyk) [inline homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx:90]: The same here.

**Review by VitalikMelnyk — CHANGES_REQUESTED** [review]
Hello @usmonbek-umarbekovich!
I have checked your PR and left some comments.

File: homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx (hunk 1 above)
Comment (usmonbek-umarbekovich) [reply homeworks/usmonbek-rustamov/18-react-ui/src/components/AvatarList.tsx]: Thank you, I learnt Mapped types because of your recommendation

**Comment by usmonbek-umarbekovich** [comment]
I handled the errors and show those errors to the user using react-toastify. I also restructured the code, seperated the code to new small components

**Review by VitalikMelnyk — APPROVED** [review]
Approved.

**Pull request merged**